tokio/src/sync/notify.rs

```diff
@@ -0,0 +1,496 @@
+}
+
+/// Initial "idle" state
+const EMPTY: u8 = 0;
+
+/// One or more threads are currently waiting to be notified.
+const WAITING: u8 = 1;
+
+/// Pending notification
+const NOTIFIED: u8 = 2;
+
+impl Notify {
+    /// Create a new `Notify`, initialized without a permit.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use tokio::sync::Notify;
+    ///
+    /// let notify = Notify::new();
+    /// ```
+    pub fn new() -> Notify {
+        Notify {
+            state: AtomicU8::new(0),
+            waiters: Mutex::new(LinkedList::new()),
+        }
+    }
+
+    /// Wait for a notification.
+    ///
+    /// Each `Notify` value holds a single permit. If a permit is available from
+    /// an earlier call to `notify_one`, then `recv` will complete immediately,
+    /// consuming that permit. Otherwise, `recv()` waits for a permit to be made
+    /// available by the next call to `notify_one`
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use tokio::sync::Notify;
+    /// use std::sync::Arc;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     let notify = Arc::new(Notify::new());
+    ///     let notify2 = notify.clone();
+    ///
+    ///     tokio::spawn(async move {
+    ///         notify2.recv().await;
+    ///         println!("received a notification");
+    ///     });
+    ///
+    ///     notify.notify_one();
+    /// }
+    /// ```
+    pub async fn recv(&self) {
+        RecvFuture {
+            notify: self,
+            state: RecvState::Init,
+            waiter: linked_list::Entry::new(Waiter {
+                waker: None,
+                notified: false,
+            }),
+        }
+        .await
+    }
+
+    /// Notifies a waiting task
+    ///
+    /// If a task is currently waiting, that task is notified. Otherwise, a
+    /// permit is stored in this `Notify` value and the **next** call to
+    /// `recv()` will complete immediately consuming the permit made available
+    /// by this call to `notify_one()`.
+    ///
+    /// At most one permit may be stored by `Notify`. Many sequential calls to
+    /// `notify_one` will result in a single permit being stored. The next call
+    /// to `recv()` will complete immediately, but the one after that will wait.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use tokio::sync::Notify;
+    /// use std::sync::Arc;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     let notify = Arc::new(Notify::new());
+    ///     let notify2 = notify.clone();
+    ///
+    ///     tokio::spawn(async move {
+    ///         notify2.recv().await;
+    ///         println!("received a notification");
+    ///     });
+    ///
+    ///     notify.notify_one();
+    /// }
+    /// ```
+    pub fn notify_one(&self) {
+        // Load the current state
+        let mut curr = self.state.load(SeqCst);
+
+        // If the state is `EMPTY`, transition to `NOTIFIED` and return.
+        while let EMPTY | NOTIFIED = curr {
```

Comment:
I don't think that's the same? The original will match if `curr` is _either_ `EMPTY` _or_ `NOTIFIED`, whereas your will match if `curr` is equal to the binary or of `EMPTY` and `NOTIFIED`. At least unless I got my precedence rules messed up, the `|` in the `let` version will be interpreted as a pattern separator, not bitwise-or.

Reply:
because clippy?

Reply:
I think @jonhoo is right. The equivalent would be:

```rust
while curr == EMPTY || curr == NOTIFIED
```` 

It's probalby a bit more natural since an or pattern is not something one sees often (I haven't seen it before). But all versions should be ok.

Reply:
@jonhoo oh, whoa, you're right — I forgot that `|` has a different meaning in patterns & interpreted it as binary or. IMO we should probably change this to 
```rust
while curr == EMPTY || curr == NOTIFIED
```
so that it's not misinterpreted by readers as `EMPTY.bit_or(NOTIFIED)`...